tests/components/hydrawise/test_entity_availability.py

```diff
@@ -0,0 +1,79 @@
+"""Test entity availability."""
+
+from collections.abc import Awaitable, Callable
+from datetime import timedelta
+from unittest.mock import AsyncMock, patch
+
+from aiohttp import ClientError
+from freezegun.api import FrozenDateTimeFactory
+from pydrawise.schema import Controller
+import pytest
+from syrupy.assertion import SnapshotAssertion
+
+from homeassistant.components.hydrawise.const import SCAN_INTERVAL
+from homeassistant.const import Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers import entity_registry as er
+
+from tests.common import MockConfigEntry, async_fire_time_changed, snapshot_platform
+
+
+@pytest.mark.parametrize(
+    ("platform"),
+    [
+        Platform.BINARY_SENSOR,
+        Platform.SWITCH,
+        Platform.SENSOR,
+    ],
+)
+async def test_controller_offline(
+    hass: HomeAssistant,
+    mock_add_config_entry: Callable[[], Awaitable[MockConfigEntry]],
+    entity_registry: er.EntityRegistry,
+    controller: Controller,
+    freezer: FrozenDateTimeFactory,
+    snapshot: SnapshotAssertion,
+    platform: Platform,
+) -> None:
+    """Test availability for sensors when controller is offline."""
+    with patch(
+        "homeassistant.components.hydrawise.PLATFORMS",
+        [platform],
+    ):
+        config_entry = await mock_add_config_entry()
+        controller.online = False
+        freezer.tick(SCAN_INTERVAL + timedelta(seconds=30))
+        async_fire_time_changed(hass)
+        await hass.async_block_till_done()
+        await snapshot_platform(hass, entity_registry, snapshot, config_entry.entry_id)
```

Comment:
Can we instead only assert that the entity is unavailable?
The other tests are already snapshotting the platform, so there is no need for it.

Reply:
> Can we instead only assert that the entity is unavailable? The other tests are already snapshotting the platform, so there is no need for it.

That seems more complicated. We would need to either hard-code the list of entities (binary_sensors, sensors, switchs) to test or enumerate them by accessing internals.  And we'd need to code the special case for the always available binary sensor.  If new sensors get added we'd need to update the hard coded list.

I suppose we could test one binary_sensor, one switch and one sensor to make sure each domain is reporting availability correctly?

Reply:
> enumerate them by accessing internals.

That's not true, as we can access the entities the same way `snapshot_platform` does.

> If new sensors get added we'd need to update the hard coded list.

This is also not true as we only have a list for special entities, which should always be available. This is fine as we are testing exactly for this.

The difference between your snapshot tests and the tests below is that we improved the readability by quickly understanding and verifying that the test is correct. The snapshot file has more than 1000 lines, there it is hard to verify that entities have the correct state.

Reply:
Thank you for your review and patience.  

> The snapshot file has more than 1000 lines, there it is hard to verify that entities have the correct state.

Excellent point, i found myself writing a script to parse the ambr to make sure it was right.

Reply:
For that reason I prefer normal test over snapshot if we assert only a single value :)